internal/utils/utils.go

```diff
@@ -0,0 +1,363 @@
+package utils
```

Comment:
Suggest to move the `utils.go` into the existing package `internal/controller`. Usually a dedicate package should be agnostic to the business logic, but I see that it's almost tightly coupled with the etcd operator's reconciliation logic.